### PR TITLE
Allow class method pass registration

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

As title.

# What issue(s) does this change relate to?

[CO-980](https://mosaicml.atlassian.net/browse/CO-980)

### CODE REVIEW

**Review by dakinggg — COMMENTED** [review]
So I realized this might not actually work, depending on how we want to allow it to be used. In the setup in this PR, if you instantiate two trainers, and register a pass on the engine of one of them, the pass also gets registered for the engine of the other one. Is that ok?

File: tests/test_passes.py [hunk 1]
```diff
@@ -40,11 +41,13 @@ def insert_dummy_algorithm(algorithms, event):
         return algorithms
 
     engine = Engine(dummy_state, dummy_logger)
+    original_passes = copy.deepcopy(engine.algorithm_passes)
     engine.register_pass(insert_dummy_algorithm)
```
Comment (dakinggg) [inline tests/test_passes.py:45]: Can you test actually calling `Engine.register_pass` on the class as well?

**Review by hanlint — COMMENTED** [review]
I was thinking about @dakinggg 's comment, and while I agree that its undesirable for passes to be shared across all engines (logically the user is intending to register passes for a particular trainer instance.), that is a rare case..

On the other hand, if users can only register passes with:
```
trainer = Trainer(...)
trainer.engine.register_pass(...)
```
then the user can never influence algorithms that run during `INIT`. 

An alternative is to allow:
```
trainer = Trainer(
    ...,
    algorithm_passes=...
)
```

And also,
```
trainer.register_passes(...)
```

File: tests/test_passes.py [hunk 2]
```diff
@@ -40,11 +41,13 @@ def insert_dummy_algorithm(algorithms, event):
...
 
     trace = engine.run_event(Event.INIT)
 
     assert 'dummy' in [tr.exit_code for tr in trace.values()]
+    Engine.algorithm_passes = original_passes
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply tests/test_passes.py:50]: oh jk, ignore this comment.

**Comment by dakinggg** [comment]
Yeah, I agree that another trainer arg seems to be the least bad option. Could also do something complicated with separate class level passes vs instance level passes, but that seems worse than adding another trainer arg.

**Comment by mvpatel2000** [comment]
Closing in favor of https://github.com/mosaicml/composer/pull/1673

**Pull request closed without merging**